Support for the Rust's custom allocator API on nightly

TeleportAura opened the pull request:
This makes it possible to use this crate with a [custom allocator](https://github.com/rust-lang/rust/issues/32838) on Rust nightly.
Imo, this should be merged because it would make this crate the first bitvec crate to be able to be used in bare metal projects

Due to the fact that custom allocators currently only work on nightly, there is some code duplication which I could probably remove using macros if this is an issue. 
The explicit implementation of PartialEq is required because the derive macro doesn't work unless the traitbound for the allocator also requires it to implement PartialEq which is a pretty stupid requirement.
Other than that, I tried to adhere to the code style in the original

I also updated the edition to be 2021 and fixed one compiler warning in a separate commit

TeleportAura commented:
This PR now also includes my commit adding no-std support

TeleportAura commented:
This PR now also includes my commit adding no-std support

marcianx commented on src/lib.rs:
```diff
@@ -9,21 +12,68 @@
+    vec: Vec<u8, A>
 }
 
+// explicitly allow comparisons between BitVecs regardless of whether
```
Super-nit: Like other doc comments, please start with capital and end with period.

marcianx commented on src/lib.rs:
```diff
@@ -9,21 +12,68 @@
+#[derive(Clone, Default)]
+pub struct BitVec<A: Allocator = Global> where Vec<u8, A>: Default {
+    nbits: usize,
+    vec: Vec<u8, A>
```
Nit: Please add trailing comma.

marcianx commented on Cargo.toml:
```diff
@@ -11,6 +12,11 @@ readme = "README.md"
+[features]
+default = ["std"]
+std = []
+custom-allocator = []
```
I would appreciate if there were some indication in the feature name that it requires unstable. E.g. serde seems to organize things like so: https://github.com/serde-rs/serde/blob/f583401284dca33d8961ed1625154d035e6d2bfb/serde/Cargo.toml#L50. You're welcome to do another variation as long as it's obvious to users of the feature that this could break at any time.

marcianx commented:
Thanks so much for your contribution!

Firstly, the main concern I have is that per its thread, the unstable feature is not in a desirable form. Does that not mean that it is very unlikely for the current state of the world to survive?

Assuming that that's resolved, then also:
1. Could you possibly add yourself to authors like I added another author in https://github.com/marcianx/downcast-rs/commit/c7d1f79cada89bd348ef8be850680d1e998fb298 and also add a contributors section as in that PR?
2. One concern I have is the ease with which one can break this functionality with a set of combinations that we can run locally. Would you be willing to add the appropriate GitHub actions if you're comfortable with them? It's fine if you're not, and I can look into afterward before publishing - but I can't get to it immediately as I'm a bit swamped currently.

TeleportAura commented:
I addressed all the issues you mentioned in the comments above in the first of my two latest commits.

Regarding the unstability of that feature, I highly doubt that the parts of the API that are important here will change any more since everything I've touched has afaik been that way since late 2016 .

1. Fixed in my latest commit
2. I sadly know next to nothing about GH actions and I need to study for a test tomorrow so I won't be able to add those

TeleportAura commented:
Whilst trying to at least advance my personal project (where I'm using this crate) by a few lines today, I realized that I had forgotten the most important thing: 
Adding a constructor for allocators that are not the global allocator

I'm extremely sorry for trying to merge what was essentially useless glue code up until that point without even noticing it.

TeleportAura commented:
Turns out that my testing was pretty flawed from the beginning. I'll get back to you either tomorrow or the day after that.
Sorry for pretty much wasting your time on this up until now.

marcianx commented:
Oh no worries! It wasn't much reviewing effort on my part, and thanks for putting all the effort you did on this.

Indeed, my concern was that without tests actually using the functionally, I couldn't confirm that it worked in the intended additional use-cases. So thanks for trying it out to discover the parts still missing. Ideally, if live too have some check - perhaps even a couple integration tests, one for each combination of supported scenarios (choice of allocator, std/no-std) - that construct the vector and do basic operations just to confirm that it works end to end. That might be sufficient to make me feel comfortable about the completeness and avoiding regressions.

TeleportAura commented:
I think I finally did it.

In regards to the testing, I added a new test for custom allocators and also made sure to run all of the tests from time to time (and before pushing the two latest commits) with the following combinations:
stable toolchain, no extra features enabled
nightly toolchain, no extra features enabled
nightly toolchain, unstable feature enabled
Sadly, this requires three separate test runs, although I do believe that an automated testing setup could handle that.
(And since nightly with unstable features covers both using a custom allocator and not using one, it is probably enough to do just that one test run)

The std feature is obsolete now as I changed every use of std to either be dependent on core alone or on alloc (which can both be built on any target supported by LLVM)

Onto the actual changes then:
I tried my best not to increase the code complexity and thus maintainability too much. However, I could only really choose between two (three) evils: Duplicating code, using macros and, last, not providing support for a feature when using custom allocators. I think that the compromise of having a reasonable set of features when using custom allocators and using macros instead of duplicating large parts of code is, although not ideal, justified.
Rust's standard library, in comparison, duplicates *HUGE* amounts of code e.g. in [Vec](https://doc.rust-lang.org/src/alloc/vec/mod.rs.html#596)'s inherent implementations which I can't imagine is very good for maintainability.
Once custom allocators are stable, I'll do some clean up and remove the then unnecessary code duplication and macros which should yield a pretty straightforward implementation.

marcianx commented on src/lib.rs:
```diff
@@ -9,21 +12,70 @@
+
+// Explicitly allow comparisons between BitVecs regardless of whether
+// they use the same allocator or whether their allocator implements
+// PartialEq or not
```
nit: please end with a period

TeleportAura replied:
is resolved

marcianx commented on src/lib.rs:
```diff
@@ -9,21 +12,70 @@
+        self.nbits == other.nbits && self.vec == other.vec
+    }
+
+    fn ne(&self, other: &BitVec<B>) -> bool {
```
optional nit: Since this is already automatically implemented, I don't see any benefit to explicitly providing it.

TeleportAura replied:
is resolved

marcianx commented on src/lib.rs:
```diff
@@ -494,7 +682,7 @@ mod test {
 
     #[test]
     fn test_convert_from_bools() {
```
I'm seeing compilation errors because `Vec` and `vec!` are not defined in tests. Please add the relevant imports so that the tests run in all 3 cases (stable, nightly, nightly+unstable).

TeleportAura replied:
That's fixed as well although I'm unsure as to why I didn't get those before

marcianx commented:
Thanks for all your effort. The code overall looks good, but please fix the compilation errors noted below.

I really wish GitHub supported something like sequential PRs. Then you could have sent me PRs for the basic clean-ups which would be trivial to review and then we could have concentrated all the attention on just the additional feature.

**Edit:** I mean, you're still welcome to split off such a PR and I can review and approve that separately and you can rebase this one on it. Would be better history too. Completely up to you.

TeleportAura commented:
everything should be fixed by the two latest commits

marcianx commented:
Thanks so much @TeleportAura!
Is this something you need pushed out as a new version?
